Test/comp/NativeNumbers.dfy

```diff
@@ -0,0 +1,54 @@
+// Skip JavaScript because JavaScript doesn't have the same native types
+
+// RUN: %dafny /compile:3 /spillTargetCode:2 /compileTarget:cs "%s" > "%t"
+// RUN: %dafny /compile:3 /spillTargetCode:2 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /compile:3 /spillTargetCode:2 /compileTarget:java "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+method Main() {
+    Int8Test();
+    Int16Test();
+}
+
+// test handling of byte/short arithmetic in Java
+newtype {:nativeType "sbyte"} int8 = x | -0x80 <= x < 0x80
+newtype {:nativeType "short"} int16 = x | -0x8000 <= x < 0x8000
+
+method Int8Test() {
```

Comment:
Could we also test the boundaries here, in particular specify what should happen when a/b are ~2^8?

Reply:
Well, these are signed integers; is it still valuable to pass in, say, 126 and 127?

Reply:
Not sure what's the most valuable.  I was just thinking I'd like to see the number wrap around, e.g. adding 1 to 2^7-1 should yield -2^7.  Or maybe Dafny doesn't allow wrapping like that because the verifier will give an error and it won't make it to the compiler?  (Forgive my lack of Dafny knowledge.)  I just don't yet see how this test is testing the type cast in Java.

Reply:
Oh—right, only casts that are known to be safe (that is, in bounds) will make it past the verifier.  These tests are just meant to make sure that code generation is keeping everything straight: declaring the right types, using the right operators, etc.

Reply:
Integer types never wrap in Dafny, and any `as` type cast is allowed only if the source expression already has a value in the target type.

Bitvector types in Dafny also have arithmetic operations, and they wrap. All bitvector values are unsigned. If this is useful for the bug that was fixed (or just as additional tests), then you may consider tests along the lines of

    var a: bv8 := 250;
    a := a + 6;
    assert a == 0;

    var b: bv7 := 126;
    b := b + 5;
    assert b == 3;

    print a, " ", b, "\n";

  Rustan

Reply:
Are bitvectors represented in Java as small types like `short` if the vector is small enough, or do we make a Java array?  I thought it was the later, but If the former, that would be a cool test.  
@lukemaurer I'm fine with this test as it is now if we can't test with bits.

Reply:
Small bitvectors get represented as primitives wrapped in our special unsigned wrapper class, which means they aren't affected by this patch.  Could be useful to have the tests for if that changes, though.

Reply:
The C# version uses unsigned integers of built-in lengths, if possible, and uses something bigger (`BigInteger`, I seem to recall) for bitvectors of more than 64 bits. From what Luke is saying, it sounds like the Java version attempts to do something similar, but uses the special unsigned wrapper classes. If that needs to change, let's do that as a separate PR (in due time).